### PR TITLE
add encrypting config

### PR DESCRIPTION
**Pull request by jkhelil**
This PR is intended to enable easily data encryption on openshift cluster regarding this documentation
https://docs.openshift.com/container-platform/3.7/admin_guide/encrypting_data.html

### CODE REVIEW

**Comment by jkhelil** [comment]
hi @michaelgugino  this PR is intended to enable data encryption related to doc here
https://docs.openshift.com/container-platform/3.7/admin_guide/encrypting_data.html
dont understand why the commit message is not clear

File: roles/openshift_control_plane/tasks/encrypting_config.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  - name: Edit master configuration
+    yedit:
+      src: "{{ openshift.common.config_base }}/master/master-config.yaml"
+      key: "kubernetesMasterConfig.apiServerArguments.experimental-encryption-provider-config"
+      value:
+      - "{{ openshift.common.config_base }}/master/encryption-config.yaml"
+      backup: true
+    register: yedit_output
+
+  - name: restart master api
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/encrypting_config.yml]: Master does not run as systemd service in 3.10+.

File: roles/openshift_control_plane/tasks/encrypting_config.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    systemd: name=atomic-openshift-master-api state=restarted
+    when: yedit_output.changed or new_config.changed
+
+  - name: ensure all secret are encrypted
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/encrypting_config.yml]: Master won't be alive at this point, can't test this here.

File: roles/openshift_control_plane/tasks/encrypting_config.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    command: >
+      {{ openshift.common.client_binary }} adm migrate storage --include=secrets --confirm
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/encrypting_config.yml]: openshift_client_binary is the variable now, also need to specify absolute path of kubeadm config file (see other usages of openshift_client_binary)

File: roles/openshift_control_plane/tasks/encrypting_config.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    register: temp_file
+
+  - name: copy new config to temp file
+    copy:
+      content: "{{ openshift_encryption_config }}"
+      dest: "{{ temp_file.path }}"
+
+  - name: stat the temporary file
+    stat:
+      path: "{{ temp_file.path }}"
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/encrypting_config.yml]: This tempfile doesn't seem to actually be copied into place.

File: roles/openshift_control_plane/tasks/encrypting_config.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    register: stat_new_encrypting_config
+
+  - name: Generate encryption config
+    copy:
+      content: "{{ openshift_encryption_config }}"
+      dest: "{{ openshift.common.config_base }}/master/encryption-config.yaml"
+    when: not stat_encrypting_file.stat.exists or stat_encrypting_file.stat.md5 != stat_new_encrypting_config.stat.md5
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/encrypting_config.yml]: We don't need to check for any of this stuff.  This should only affect new installs, just copy the data into place.

File: roles/openshift_control_plane/tasks/encrypting_config.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+---
+- when: openshift_encryption_config is defined
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/encrypting_config.yml]: put this when condition on the import_tasks instead of making this whole file a block.

**Comment by rocky4570** [comment]
Can just set your master encryption file (for now in openshift inventory file) via 

osm_api_server_args={"experimental-encryption-provider-config": ["{{ etcd_encryption_file }}"]}

Then your ansible can create the file as per the Openshift doco to setup the encryption-config using
etcd_encryption_file: "/etc/origin/master/encryption-config.yml"

its annoying that its still called experimental-encryption-provider-config when it seems to have been available from 3.7. We got this working right at the start of the playbook so you put the encryption-config.yml in place and the build process will begin using it.

File: roles/openshift_control_plane/tasks/main.yml [hunk 7]
```diff
@@ -65,6 +65,12 @@
     dest: "{{ openshift_master_scheduler_conf }}"
     backup: true
 
+- name: Generate encryption config
+  copy:
+    content: "{{ openshift_encryption_config }}"
+    dest: "{{ openshift.common.config_base }}/master/encryption-config.yaml"
+  when: api_server_args is defined and api_server_args['experimental-encryption-provider-config'] is defined and openshift_encryption_config is defined
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/main.yml]: api_server_args does not appear to be a valid variable name.  In any case, we don't need to check for 'api_server_args' in any form, we can just put this file if openshfit_encryption_config is defined, master's config will be handled independently of this task.

Comment (rocky4570) [reply roles/openshift_control_plane/tasks/main.yml]: I found it was osm_api_server_args as in the above posting this works with 3.10.45 recently.

File: roles/openshift_control_plane/tasks/main.yml [hunk 8]
```diff
@@ -66,6 +66,12 @@
     dest: "{{ openshift_master_scheduler_conf }}"
     backup: true
 
+- name: Generate encryption config
+  copy:
+    content: "{{ openshift_encryption_config }}"
+    dest: "{{ openshift.common.config_base }}/master/encryption-config.yaml"
+  when: openshift_encryption_config is defined
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/main.yml]: Lint failed:
roles/openshift_control_plane/tasks/main.yml:
  73:47     error    trailing spaces  (trailing-spaces)

File: inventory/hosts.example [hunk 9]
```diff
@@ -341,7 +341,10 @@ debug_level=2
 #osm_controller_args={'resource-quota-sync-period': ['10s']}
 
 # Configure api server arguments
-#osm_api_server_args={'max-requests-inflight': ['400']}
+#osm_api_server_args={'max-requests-inflight': ['400'],'experimental-encryption-provider-config':['/etc/origin/master/encryption-config.yaml']}
+
+# Configure openshift encrypting config
+#openshift_encryption_config="{'kind': 'EncryptionConfig','apiVersion': 'v1','resources': [{'resources': ['secrets'],'providers': [{'aescbc': {'keys': [{'name': 'key1','secret': 'xxxxxxxx'}]}},{'identity': {}}]}]}"
```
Comment (michaelgugino) [inline inventory/hosts.example]: What is the desired format of this data?  Currently, this is a python dictionary serialized into a string.  If you are looking to have json here, need to use double quotes for elements and single quotes to wrap the entire value.

**Pull request closed without merging**